tests/lib/ocpp/v201/test_smart_charging_handler.cpp

```diff
@@ -122,36 +145,40 @@ class ChargepointTestFixtureV201 : public testing::Test {
                                .chargingSchedule = charging_schedules,
                                .customData = {},
                                .recurrencyKind = recurrency_kind,
-                               .validFrom = {},
-                               .validTo = {},
+                               .validFrom = validFrom,
+                               .validTo = validTo,
                                .transactionId = transaction_id};
     }
 
-    ChargingProfile create_tx_profile_with_missing_transaction_id(ChargingSchedule charging_schedule) {
-        auto charging_profile_id = DEFAULT_PROFILE_ID;
-        auto stack_level = DEFAULT_STACK_LEVEL;
-        auto charging_profile_purpose = ChargingProfilePurposeEnum::TxProfile;
-        auto charging_profile_kind = ChargingProfileKindEnum::Absolute;
-        auto recurrency_kind = RecurrencyKindEnum::Daily;
-        std::vector<ChargingSchedule> charging_schedules = {charging_schedule};
-        return ChargingProfile{
-            charging_profile_id,
-            stack_level,
-            charging_profile_purpose,
-            charging_profile_kind,
-            charging_schedules,
-            {}, // transactionId
-            recurrency_kind,
-            {}, // validFrom
-            {}  // validTo
-        };
+    void create_device_model_db(const std::string& path) {
+        sqlite3* source_handle;
+        sqlite3_open(DEVICE_MODEL_DB_LOCATION_V201, &source_handle);
+        sqlite3_open(path.c_str(), &db_handle);
+
+        auto* backup = sqlite3_backup_init(db_handle, "main", source_handle, "main");
+        sqlite3_backup_step(backup, -1);
+        sqlite3_backup_finish(backup);
+
+        sqlite3_close(source_handle);
     }
 
-    DeviceModel create_device_model() {
-        std::unique_ptr<DeviceModelStorageMock> storage_mock =
-            std::make_unique<testing::NiceMock<DeviceModelStorageMock>>();
-        ON_CALL(*storage_mock, get_device_model).WillByDefault(testing::Return(DeviceModelMap()));
-        return DeviceModel(std::move(storage_mock));
+    std::shared_ptr<DeviceModel>
+    create_device_model(const std::string& path = "file:device_model?mode=memory&cache=shared",
+                        const std::optional<std::string> ac_phase_switching_supported = "true") {
```

Comment:
This method sets the defaults so that all of the other test cases pass. For the ACPhaseSwitchingSupported tests, we need to do this as we can't make the value blank after it's already been initialized with a valid value. So we can't do what we do with e.g. the FR26 tests.

As written, this function allows us to initialize a device model with defaults that will work for all cases, including the case where we want to test our code for when the value is missing.

Reply:
To be more explicit, my feedback is that `create_device_model` should take in a single argument, which is the path. It should use defaults for everything else. Each test should then override the parts of the device model that are relevant to itself.

Concretely, I don't see why `ac_phase_switching_supported` is special enough that it needs to be an argument but (say) `charging_rate_unit_cv.component` is not.